daemon/containerd/image_tag.go

```diff
@@ -1,20 +1,57 @@
 package containerd
 
 import (
-	"errors"
+	"context"
 
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	containerdimages "github.com/containerd/containerd/images"
 	"github.com/docker/distribution/reference"
 	"github.com/docker/docker/errdefs"
 	"github.com/docker/docker/image"
+	"github.com/sirupsen/logrus"
 )
 
-// TagImage creates the tag specified by newTag, pointing to the image named
-// imageName (alternatively, imageName can also be an image ID).
-func (i *ImageService) TagImage(imageName, repository, tag string) (string, error) {
-	return "", errdefs.NotImplemented(errors.New("not implemented"))
-}
+// TagImageWithReference creates an image named as newTag and targetting the given descriptor id.
+func (i *ImageService) TagImageWithReference(ctx context.Context, imageID image.ID, newTag reference.Named) error {
+	logger := logrus.WithFields(logrus.Fields{
+		"imageID": imageID.String(),
+		"tag":     newTag.String(),
+	})
+
+	target, err := i.resolveDescriptor(ctx, imageID.String())
+	if err != nil {
+		logger.WithError(err).Debug("failed to resolve image id to a descriptor")
+		return err
+	}
+
+	img := containerdimages.Image{
+		Name:   newTag.String(),
+		Target: target,
+	}
+
+	is := i.client.ImageService()
+	_, err = is.Create(ctx, img)
+	if err != nil {
+		if !cerrdefs.IsAlreadyExists(err) {
+			logger.WithError(err).Debug("failed to create image")
+			return errdefs.System(err)
+		}
+
+		// If there already exists an image with this tag, delete it
+		err := is.Delete(ctx, img.Name)
+
+		if err != nil {
+			logger.WithError(err).Error("failed to delete old image")
```

Comment:
The other logs are `Debug` these could be too no?

Reply:
My reasoning was that the first one can be Debug, because it mostly should be just an user error (non existing id) and it's kind of expected to happen normally.

But if the image couldn't be created and it's not an "already exists" error, then probably it's something that's not expected to happen under normal circumstances s and it's worth having logging without `debug` enabled.

Reply:
> But if the image couldn't be created and it's not an "already exists" error, then probably it's something that's not expected to happen under normal circumstances s and it's worth having logging without debug enabled.

That one should probably be an `Error()`, as it's an unexpected error (and we currently return a `errdefs.System()` (internal server error).

Reply:
Really we shouldn't need to log and return an error.

Reply:
The log has additional information about the context (imageID, tag) - which might be useful when debugging failures. The only caveat is that these will only appear in the daemon log.
Should we drop the logging and just include these in the returned error?

Reply:
I think it makes sense to include it in the error.

Reply:
Removed the extra `logrus` logs and included details directly in returned errors.